### ISSUE TITLE
Client get stucked while writing

### ISSUE THREAD

**Issue by lynus** [Alluxio/alluxio#14811]
**Alluxio Version:**
What version of Alluxio are you using?
2.7.2 and git tip
**Describe the bug**
client always get stucked while writing files.
**To Reproduce**
Steps to reproduce the behavior (as minimally and precisely as possible)
 ./bin/alluxio runClass alluxio.stress.cli.client.StressClientIOBench --operation write --file-size 25g --base alluxio://11.22.33.51:19998/client_stress  --threads 8
or
 ./bin/alluxio runClass alluxio.stress.cli.client.StressClientIOBench --operation write --file-size 100g --base alluxio://11.22.33.51:19998/client_stress  --threads 1
**Expected behavior**
The stress test should success.
**Urgency**
Describe the impact and urgency of the bug.
no sure 
**Are you planning to fix it**
Please indicate if you are already working on a PR. 
I'm working on it.
**Additional context**
no

**Comment by lynus** [comment]
the stucking thread dump:
Thread "bench-thread-0":
 at sun.misc.Unsafe.park(boolean, long)
 at java.util.concurrent.locks.LockSupport.parkNanos(java.lang.Object, long) (line: 215)
 at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(long, java.util.concurrent.TimeUnit) (line: 2163)
 at alluxio.resource.DynamicResourcePool.acquire(long, java.util.concurrent.TimeUnit) (line: 350)
 at alluxio.resource.DynamicResourcePool.acquire() (line: 297)
 at alluxio.client.file.FileSystemContext.acquireBlockWorkerClientInternal(alluxio.wire.WorkerNetAddress, alluxio.ClientContext, alluxio.security.user.UserState) (line: 549)
 at alluxio.client.file.FileSystemContext.acquireBlockWorkerClient(alluxio.wire.WorkerNetAddress) (line: 529)
 at alluxio.client.block.stream.LocalFileDataWriter.create(alluxio.client.file.FileSystemContext, alluxio.wire.WorkerNetAddress, long, long, alluxio.client.file.options.OutStreamOptions) (line: 79)
 at alluxio.client.block.stream.DataWriter$Factory.create(alluxio.client.file.FileSystemContext, long, long, alluxio.wire.WorkerNetAddress, alluxio.client.file.options.OutStreamOptions) (line: 87)
 at alluxio.client.block.AlluxioBlockStore.getOutStream(long, long, alluxio.wire.WorkerNetAddress, alluxio.client.file.options.OutStreamOptions) (line: 292)
 at alluxio.client.block.AlluxioBlockStore.getOutStream(long, long, alluxio.client.file.options.OutStreamOptions) (line: 331)
 at alluxio.client.file.AlluxioFileOutStream.getNextBlock() (line: 302)
 at alluxio.client.file.AlluxioFileOutStream.writeInternal(byte[ ], int, int) (line: 268)
 at alluxio.client.file.AlluxioFileOutStream.write(byte[ ], int, int) (line: 235)
 at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.write(byte[ ], int, int) (line: 57)
 at java.io.DataOutputStream.write(byte[ ], int, int) (line: 107)
 at alluxio.stress.cli.client.StressClientIOBench$AlluxioHDFSBenchThread.applyOperation() (line: 497)
 at alluxio.stress.cli.client.StressClientIOBench$BenchThread.runInternal() (line: 402)
 at alluxio.stress.cli.client.StressClientIOBench$BenchThread.call() (line: 370)
 at alluxio.stress.cli.client.StressClientIOBench$BenchThread.call() (line: 328)
 at java.util.concurrent.FutureTask.run() (line: 266)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (line: 1149)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run() (line: 624)
 at java.lang.Thread.run() (line: 748)

**Comment by lynus** [comment]
And this is the error report from the stress benchmark:    "errors" : [ "Failed to cache: Thread interrupted while acquiring client from pool: alluxio.client.block.stream.BlockWorkerClientPool@6ec49b4" ]

**Comment by lynus** [comment]
I think I found the cause. the AlluxioFileOutputStream opens a writer resource for each block. A writer resource is held until AlluxioFileOutpurtStream closes. So when  app writing  a really big file, it would exhaustes the writer pool because during writing no resources get chances to  return to the pool.

**Comment by yuzhu** [comment]
@jiacheliu3 have you encountered this writer pool issue before?

**Comment by jiacheliu3** [comment]
> I think I found the cause. the AlluxioFileOutputStream opens a writer resource for each block. A writer resource is held until AlluxioFileOutpurtStream closes. So when app writing a really big file, it would exhaustes the writer pool because during writing no resources get chances to return to the pool.

Yes this is a potential improvement tracked here https://github.com/Alluxio/alluxio/issues/14798

Good job figuring this out! You can tune up `alluxio.user.block.worker.client.pool.max` to get around this, essentially by increasing the concurrency limit. 

@beinan This blocking behavior when writing large files (with concurrency lower than 1024) is an issue, because `alluxio.user.block.worker.client.pool.max` is no longer giving us the concurrency control. The concurrency when writing large files is much lower now! One more reason why the `alluxio.user.block.worker.client.pool.max` should be re-designed.
